posix/unistd.c

```diff
@@ -0,0 +1,609 @@
+
+    oe_strlcpy(name, uts.domainname, len);
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+static char _cwd[OE_PATH_MAX] = "/";
+static oe_pthread_spinlock_t _lock;
+
+char* oe_getcwd(char* buf, size_t size)
+{
+    char* ret = NULL;
+    char* p = NULL;
+    size_t n;
+    bool locked = false;
+
+    if (buf && size == 0)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (!buf)
+    {
+        n = OE_PATH_MAX;
+        p = oe_malloc(n);
+
+        if (!p)
+            OE_RAISE_ERRNO(OE_ENOMEM);
+    }
+    else
+    {
+        n = size;
+        p = buf;
+    }
+
+    oe_pthread_spin_lock(&_lock);
+    locked = true;
+
+    if (oe_strlcpy(p, _cwd, n) >= n)
+        OE_RAISE_ERRNO(OE_ERANGE);
+
+    oe_pthread_spin_unlock(&_lock);
+    locked = false;
+
+    ret = p;
+    p = NULL;
+
+done:
+
+    if (locked)
+        oe_pthread_spin_unlock(&_lock);
+
+    if (p && p != buf)
+        oe_free(p);
+
+    return ret;
+}
+
+int oe_chdir(const char* path)
+{
+    int ret = -1;
+    char real_path[OE_PATH_MAX];
+    struct oe_stat st;
+    bool locked = false;
+
+    /* Resolve to an absolute canonical path. */
+    if (!oe_realpath(path, real_path))
+        return -1;
+
+    /* Fail if unable to stat the path. */
+    if (oe_stat(real_path, &st) != 0)
+    {
+        // oe_errno set by oe_stat().
+        return -1;
+    }
+
+    /* Fail if path not a directory. */
+    if (!OE_S_ISDIR(st.st_mode))
+    {
+        oe_errno = OE_ENOTDIR;
+        return -1;
+    }
+
+    /* Set the _cwd global. */
+    oe_pthread_spin_lock(&_lock);
+    locked = true;
+
+    if (oe_strlcpy(_cwd, real_path, OE_PATH_MAX) >= OE_PATH_MAX)
+        OE_RAISE_ERRNO(OE_ENAMETOOLONG);
+
+    oe_pthread_spin_unlock(&_lock);
+    locked = false;
+
+    ret = 0;
+
+done:
+
+    if (locked)
+        oe_pthread_spin_unlock(&_lock);
+
+    return ret;
+}
+
+unsigned int oe_sleep(unsigned int seconds)
+{
+    const uint64_t ONE_SECOND = 1000;
+    const uint64_t msec = seconds * ONE_SECOND;
+
+    return (oe_sleep_msec(msec) == 0) ? 0 : seconds;
+}
+
+oe_pid_t oe_getpid(void)
+{
+    oe_pid_t ret = 0;
+    oe_posix_getpid(&ret);
+    return ret;
+}
+
+oe_pid_t oe_getppid(void)
+{
+    oe_pid_t ret = 0;
+    oe_posix_getppid(&ret);
+    return ret;
+}
+
+oe_pid_t oe_getpgrp(void)
+{
+    oe_pid_t ret = 0;
+    oe_posix_getpgrp(&ret);
+    return ret;
+}
+
+oe_uid_t oe_getuid(void)
+{
+    oe_uid_t ret = 0;
+    oe_posix_getuid(&ret);
+    return ret;
+}
+
+oe_uid_t oe_geteuid(void)
+{
+    oe_uid_t ret = 0;
+    oe_posix_geteuid(&ret);
+    return ret;
+}
+
+oe_gid_t oe_getgid(void)
+{
+    oe_gid_t ret = 0;
+    oe_posix_getgid(&ret);
+    return ret;
+}
+
+oe_gid_t oe_getegid(void)
+{
+    oe_gid_t ret = 0;
+    oe_posix_getegid(&ret);
+    return ret;
+}
+
+oe_pid_t oe_getpgid(oe_pid_t pid)
+{
+    oe_pid_t ret = -1;
+    oe_pid_t retval = -1;
+
+    if (oe_posix_getpgid(&retval, pid) != OE_OK)
+    {
+        oe_errno = OE_EINVAL;
+        goto done;
+    }
+
+    if (retval == -1)
```

Comment:
Do we need retval and ret here? It appears that ret is always retval before this function returns. Same for next function

Reply:
Yes, we are following a policy of not setting **ret** until just before done. Otherwise, the code is harder to follow when **ret** is set earlier and then intervening code is added (it is unlikely that the author will bother introducing an interim variable to maintain clarity).